thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobScheduler.java

```diff
@@ -264,4 +280,241 @@ private Long getIdFromJobKey(String jobKey) {
     String id = tokens[tokens.length - 1];
     return Long.valueOf(id);
   }
+
+  /**
+   * Sequentially performs anomaly detection for all the monitoring windows that are located between backfillStartTime
+   * and backfillEndTime. A lightweight job is performed right after each detection job and notified is set to false in
+   * order to silence the mail alerts.
+   *
+   * NOTE: We assume that the backfill window for the same function DOES NOT overlap. In other words, this function
+   * does not guarantees correctness of the detections result if it is invoked twice with the same parameters.
+   *
+   * @param functionId the id of the anomaly function, which has to be an active function
+   * @param backfillStartTime the start time for backfilling
+   * @param backfillEndTime the end time for backfilling
+   */
+  public void runBackfill(long functionId, DateTime backfillStartTime, DateTime backfillEndTime) {
+    AnomalyFunctionDTO anomalyFunction = anomalyFunctionDAO.findById(functionId);
+    boolean isActive = anomalyFunction.getIsActive();
+    if (!isActive) {
+      LOG.info("Skipping function {}", functionId);
+      return;
+    }
+
+    BackfillKey backfillKey = new BackfillKey(functionId, backfillStartTime, backfillEndTime);
+    Thread returnedThread = existingBackfillJobs.putIfAbsent(backfillKey, Thread.currentThread());
+    // If returned thread is not current thread, then a backfill job is already running
+    if (returnedThread != null) {
+      LOG.info("Aborting... An existing back-fill job is running...");
+      return;
+    }
+
+    try {
+      CronExpression cronExpression = null;
+      try {
+        cronExpression = new CronExpression(anomalyFunction.getCron());
+      } catch (ParseException e) {
+        LOG.error("Failed to parse cron expression for function {}", functionId);
+        return;
+      }
+
+      // Check out if there exists any previous jobs
+      DateTime currentStart;
+      JobDTO previousJob = getPreviousJob(functionId, backfillStartTime.getMillis(), backfillEndTime.getMillis());
+      if (previousJob != null) {
```

Comment:
can you add comments on what this if-block is achieving?. Also, extract it to a different method.